bigquery/iterator.go

```diff
@@ -194,6 +194,31 @@ func (it *RowIterator) fetch(pageSize int, pageToken string) (string, error) {
 	return res.pageToken, nil
 }
 
+// ArrowIterator is a raw interface for getting data from Storage Read API in arrow format
+type ArrowIterator struct {
+	// TotalRows is the estimated total number of rows in the result.
+	TotalRows uint64 
+	r         *RowIterator
+}
+
+// NextRecord returns next batch of rows as serialised arrow.Record
+func (a *ArrowIterator) NextRecord() ([]byte, error) {
+	return a.r.arrowIterator.next()
+}
+
+// Schema is available after first call to NextRecord
+func (a *ArrowIterator) Schema() []byte {
+	return a.r.arrowIterator.decoder.rawArrowSchema
+}
```

Comment:
It might be more interoperable with the Arrow libraries if you could expose a direct `io.Reader` for the raw IPC stream

Reply:
Actually I am using this to construct a `array.RecordReader`. Sample here : https://github.com/rilldata/rill/blob/09726146ec04415e68f7e41efa64e5b5db12aa7e/runtime/drivers/bigquery/arrow.go#L19
I can move this code to the SDK and make `ArrowIterator` confirm to `array.RecordReader` as proposed in other comments as well. Is the `array.RecordReader` locked for further modifications ?

It seemed tricky to expose a `io.Reader` that can read from multiple IPC streams.
Also I was looking at the spec for `io.Reader` and saw this statement : `If some data is available but not len(p) bytes, Read conventionally returns what is available instead of waiting for more.`
I am wondering what should be the behaviour here? Should we wait for next API call to succeed or return whatever is available ?

Reply:
The difficulty with having `ArrowIterator` conform to `array.RecordReader` is that it would require explicitly exposing the major versioned `arrow.Record` interface which I believe was trying to be avoided. If it's okay to directly include the versioned `arrow.Record` interface in the API then I would agree that conforming to the `array.RecordReader` interface would be best.

I agree you wouldn't expose an `io.Reader` to read from multiple IPC streams, but rather you'd expose an individual `io.Reader` for each IPC stream, and then allow the consumer to construct an `array.RecordReader` for each stream. That's the common way of doing it to allow the consumer to control the backpressure etc. It's also how I implemented the Snowflake arrow stream handling, it provides an `io.Reader` *per* stream.

Reply:
Thanks for the inputs @zeroshade 
Should the `io.Reader` implementation wait for the next batch to be available or return whatever bytes are available as per the `io.Reader` specs?

Reply:
it should return whatever bytes are available as per the `io.Reader` spec. Just make sure you don't return an `io.EOF` unless you *actually* are at the end :smile: